Autolab partial feedback endpoints

michellexliu opened the pull request:
## Description
Calls the Tango partial output and live jobs endpoints from Autolab to render more detailed job status and display partial output.

### New feedback view states
**In Progress submissions:**
![Screen Shot 2022-11-29 at 1 41 43 AM](https://user-images.githubusercontent.com/50491000/204552245-7400f938-a8ff-4d82-8e8c-e4af8f3c0f0e.jpg)
Users are able to press a refresh button to get updated partial output, which gets displayed in the feedback section. The refresh button is rate-limited from the frontend such that users may only call a refresh every 5 seconds. If there is no longer any partial output, the refresh will trigger a page reload to check for full output.

**Queued submissions:**
![Screen Shot 2022-11-29 at 1 41 11 AM](https://user-images.githubusercontent.com/50491000/204552657-ee348dc2-ca18-43d1-9c1b-043c92da3449.jpg)
If a submission is queued, a "queued" page will get displayed, as well as the student's queue position. There is similarly a rate-limited refresh button.

**Completed submissions:**
Once autograding is complete, the feedback is displayed as normal, with no refresh button.

### Notable behavior changes
When the autograder is still running, the `--` in the grades table temporarily becomes clickable. If users click into it, they are led to the view feedback page, where the possible views mentioned above are shown.
![image](https://user-images.githubusercontent.com/50491000/204981614-f3f4d118-e6cb-430a-8be3-bfafac54c8e3.png)

After the autograder finishes running, the auto-graded scores are updated to numbers and remain clickable. Non-auto-graded scores go back to the original behavior. 
![image](https://user-images.githubusercontent.com/50491000/204982212-37d96ae0-5e06-48d8-ae61-70303b5844bf.png)

If a user has clicked into the viewFeedback page of a non-auto-graded problem while the file was still being autograded, and then refreshes on it after autograding is complete, they will be redirected with a "No feedback for requested score" error.

## Motivation and Context
This change allows students to view feedback as soon as it is updated, rather than just at the end. This is especially important for classes with long test cases, where students may not want to wait until their code is completely done to be able to see individual test case results.

## How Has This Been Tested?

### Setup
1. Run Tango (`redis-server`, `python3 restful_tango/server.py`, `python jobManager.py`)
2. Created assignment with [Python autograder](https://github.com/autolab/autograders-examples/tree/main/python-twosumlab)
3. Create Python files that print every second for 30 seconds, 60 seconds, 90 seconds, etc.

### Test scenarios:
* Standard autograder feedback complete view: displayed as "complete", no refresh button, feedback loads
* Standard autograder feedback complete view with an autograder error: displayed as "complete", no refresh button, feedback loads
* Manual feedback: displayed as "complete", no refresh button, feedback loads
* In-progress submission: loads properly, refresh button works
* Queued submission: successfully shows queue position, refresh button works
* In-progress -> in-progress refresh, In-progress -> finished running, in-progress -> finished running with Autograder error, queued -> finished running with semantic config
* Queued -> updated queued position refresh, queued -> in-progress, queued -> finished running, queued -> finished running with Autograder error, queued -> finished running with semantic config
* Output with semantic configuration
* All of the above in both student view & instructor view

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have run rubocop for style check. If you haven't, run `overcommit --install && overcommit --sign` to use pre-commit hook for linting
- [ ] My change requires a change to the documentation, which is located at [Autolab Docs](https://github.com/autolab/docs)
- [ ] I have updated the documentation accordingly, included in this PR

victorhuangwq commented:
I haven't yet look at the codebase, but one suggestion I have right now is that we should now change the page that the users see after submitting from the current handin_history page, to this page if there is an autograder setup.

victorhuangwq commented on app/views/assessments/_partial_feedback_js.html.erb:
```diff
@@ -0,0 +1,65 @@
+  const btn = $('#refresh-feedback')[0];
+
+  // Throttling Function
+  const throttleFunction=(func, delay)=>{
```
We could use lodash's throttle function. https://lodash.com/docs/3.10.1#throttle 

We have already included lodash into our codebase as a dependency.  Refer to 
/Autolab/app/helpers/application_helper.rb and 
/Autolab/app/views/metrics/index.html.erb

michellexliu commented:
> I haven't yet look at the codebase, but one suggestion I have right now is that we should now change the page that the users see after submitting from the current handin_history page, to this page if there is an autograder setup.
> 
> Else it's hard for people to know that this feature exist. Because a clickable "--" doesn't signal to the user that they can look at partial feedback upon clicking. Alternatively, you can still redirect them to the handin history page, but put a banner up at the top to tell them that their latest submission is currently being autograded

@victorhuangwq How does this look? I added a link to view feedback in the flash and changed the link to a clock icon instead of `--` if it's grading.
![image](https://user-images.githubusercontent.com/50491000/205100113-34786c09-5249-450a-a1b8-9e72cd9605d9.png)

michellexliu commented on app/controllers/assessments_controller.rb:
```diff
@@ -587,6 +594,30 @@ def viewFeedback
 
+  action_auth_level :getPartialFeedback, :student
+
+  def getPartialFeedback
```
yeah i had tried getting rack-attack earlier by adding `getPartialFeedback` to the `rack-attack.rb` file, but couldn't quite get it to work :( can try to give it another go though

victorhuangwq replied:
okay we can leave it be for now, and create a new issue

victorhuangwq commented:
Yes the clock makes a lot more sense, and I think the banner is good too :)

victorhuangwq commented:
I tested using hello.c lab and submitted a random text file

<img width="804" alt="image" src="https://user-images.githubusercontent.com/5773562/205176259-e5b66349-03ee-44b4-8ce3-d0afed81a48f.png">

Although results are returned, the clock remains. Clicking into it gets me into a loop

victorhuangwq commented:
Tested using hello.c with a sleep of 2 second and fflush(stdout).
- Clicking refresh works as expected. However
<img width="241" alt="image" src="https://user-images.githubusercontent.com/5773562/205177007-5ea76605-4b76-40ab-b533-3a265a37dc74.png">, clicking on the refresh button does not give me feedback that it has been registered (no UI feedback)
- It doesn't seem to transit from in progress to completed for me automatically, unless I click refresh

victorhuangwq commented on app/views/assessments/_partial_feedback_js.html.erb:
```diff
@@ -0,0 +1,50 @@
+          } else if (!res.is_assigned) {
+            /* if there's no partial feedback and no queue position,
+                refresh the page to see if full feedback is available */
+            location.reload();
```
I think this is the line that is causing the refresh loop. Basically I think my tango is acting weird (partialfeedback endpoint was not setup properly, so it doesn't work), and thus there is no partial feedback and no queue position.
I don't think what we should do here is to reload again tho

victorhuangwq replied:
I have figured out how to replicate this bug:
- Set up Tango without the PR that introduces partial feedback
- Submit an assignment normally
- Because partial feedback isn't working (the endpoint doesn't exist), so it's confused causing this bootloop

victorhuangwq replied:
Basically, there should be error handling when the partial feedback endpoint throws an error. 
Here's what currently happens when autograding fails (other parts of it)
![image](https://user-images.githubusercontent.com/5773562/205335040-002ee284-55c5-47db-b75a-8684b0f113ac.png)

victorhuangwq commented on app/controllers/assessments_controller.rb:
```diff
@@ -587,6 +604,30 @@ def viewFeedback
+      resp["queue_length"] = @job_status["queue_length"]
+    end
+
+    render json: resp.to_json
```
So relevant to the other situation with regards to the refresh loop, you should just throw an error instead of return partial feedback

michellexliu replied:
flash[:error] & redirect were not working... added an error message from frontend to handle this instead

victorhuangwq commented on app/views/assessments/viewFeedback.html.erb:
```diff
@@ -1,120 +1,40 @@
+        <% if @score.nil? %>
+          <div class="feedback-header">
+            <h3>Feedback</h3>
+            <div id="status-refresh">
```
Like I mentioned in the other comment, the refresh button should "spin" when you click on it.
This is important because sometimes when a user refresh, but nothing is getting outputted from the course autograder. So they don't know if the click of the refresh is being registered.

For an example of the refresh button behavior, refer to the refresh button on the metrics page http://localhost:3001/courses/AutoPopulated/metrics/index

michellexliu replied:
done

damianhxy replied:
For some reason, the refresh icon at the top rotates off-center (at least, sometimes). The one at the bottom is fine though.

https://user-images.githubusercontent.com/9074856/205424604-c0352319-42a2-445f-8a20-6abf706c3aed.mov

victorhuangwq reviewed (changes requested):
Tested locally. Works well for the ideal use case.

Requested changes for:
- Spinning behavior of refresh on click
- Better error handling in the Autolab's partial feedback endpoint and in the refreshPartial function

victorhuangwq reviewed (approved):
LGTM. 
Tested locally, with both old tango and new tango (with partial output) and works as expected.

damianhxy commented:
Tangentially related:

With the removal of the `max-width: 970px` in the previous streaming output PR, I observed the following issue where the handin form is too "narrow"

![Screenshot 2022-12-02 at 23 20 55](https://user-images.githubusercontent.com/9074856/205422435-c0b931f3-575f-4450-8c83-a2fa38e4e381.png)

This is because the div has a `col` style applied without any modifier. We should add a `s12` class as follows in `assessments/_handin_form.html.erb`

![Screenshot 2022-12-02 at 23 21 18](https://user-images.githubusercontent.com/9074856/205422447-fb4be4dd-1309-477e-8dfc-584bace33833.png)

So that it renders properly

![Screenshot 2022-12-02 at 23 22 37](https://user-images.githubusercontent.com/9074856/205422646-7abad49e-832e-4914-8fb6-07c01193b22a.png)

Note: I did a grep and this seems to be the only instance of a `class="col"`

damianhxy commented on app/views/assessments/_partial_feedback_js.html.erb:
```diff
@@ -0,0 +1,66 @@
+          } else if (res.queue_position != null) {
+            $('#partial-feedback').html(`
+              <div class="autograding-in-progress">
+                <pre>You are at position ${(res.queue_position) + 1} / ${(res.queue_length)} in the autograding queue.</pre>
```
Nit: Do students need to know / care about the `queue_length`? Perhaps displaying only `queue_position` would be sufficient.

damianhxy commented on app/views/assessments/_partial_feedback_js.html.erb:
```diff
@@ -0,0 +1,66 @@
+                Your submission is currently in the autograding queue, waiting to be graded.
+              </p>
+              <p>
+                Press the refresh button to get updates.
```
Nit: `Click` is more semantically correct than `press` for a software button

damianhxy commented on app/views/assessments/_results_panel.html.erb:
```diff
@@ -0,0 +1,56 @@
+        <h4>Autograder Scores</h4>
+        <table class="ui unstackable table">
+          <tbody>
+          <% @scoreHash.each do |key, value| %>
```
Nit: Could we do something like `@scoreHash.except("_total")` and remove the `key != '_total'` check below

damianhxy commented on app/views/assessments/_results_panel.html.erb:
```diff
@@ -0,0 +1,56 @@
+            <% if key != '_total' %>
+              <tr>
+                <td><%= key %>:</td>
+                <td><%= value %></td>
```
Nit: value should minimally be rounded to 1 d.p. to be consistent with the submission summary display (`value.to_f.round(1)`)

Same with the autograding total below

damianhxy reviewed (changes requested):
Followed testing plan, semantic config works too, code generally looks good

Requesting the following changes
- Add `s12` to handin form `div`
- Update logic for `@problemReleased`
- Remove extraneous schema changes
- Address off-center rotation of refresh button

Commented some other nits too

damianhxy reviewed (approved):
Other than nits, LGTM
